core/src/main/java/com/linecorp/armeria/server/conditional/ConditionalRequestUtil.java

```diff
@@ -0,0 +1,209 @@
+                return PERFORM_METHOD;
+            }
+        }
+        if (dataETag != null) {
+            if (weakComparison(requestETags, dataETag)) {
+                return SKIP_METHOD_NOT_MODIFIED;
+            }
+        }
+        return PERFORM_METHOD;
+    }
+
+    /**
+     * Performs an If-None-Match.
+     * @param header The If-None-Match header.
+     * @param dataETag The ETag of the current data. If the resource doesn't exists, this is null.
+     *      This is mostly used to make sure PUT operations don't overwrite existing data.
+     * @return true if the method should be performed (ie. NOT just return 304 Not Modified).
+     */
+    public static ETagResponse ifNoneMatch(@Nullable String header, @Nullable ETag dataETag) {
+        final List<ETag> requestETags = parseHeader(header);
+        if (requestETags != null) {
+            return ifNoneMatch(requestETags, dataETag);
+        }
+        return PERFORM_METHOD;
+    }
+
+    /**
+     * Processes a request doing If-Match, If-None-Match, If-Unmodified-Since and If-Modified-Since.
+     * @param reqHeaders The RequestHeaders object of the request.
+     * @param dataETag The ETag of the current data. If the resource doesn't exists, this is null.
+     *      This is mostly used to make sure PUT operations don't overwrite existing data.
+     * @return HttpStatus to return. PRECONDITION_FAILED for when If-Match matches.
+     */
+    public static ETagResponse conditionalRequest(RequestHeaders reqHeaders,
+                                                  @Nullable ETag dataETag,
+                                                  @Nullable Long dataLastModified) {
+        requireNonNull(reqHeaders, "reqHeaders cannot be null");
+        final List<ETag> ifMatchHeader = parseHeader(reqHeaders.get(HttpHeaderNames.IF_MATCH));
+        if (ifMatchHeader != null && ifMatch(ifMatchHeader, dataETag) == SKIP_METHOD_PRECONDITION_FAILED) {
+            return SKIP_METHOD_PRECONDITION_FAILED;
+        }
+        final String ifNoneMatchHeader = reqHeaders.get(HttpHeaderNames.IF_NONE_MATCH);
+        if (ifNoneMatchHeader != null) {
+            final ETagResponse ifNoneMatchResponse = ifNoneMatch(ifNoneMatchHeader, dataETag);
+            if (ifNoneMatchResponse == SKIP_METHOD_NOT_MODIFIED) {
+                return SKIP_METHOD_NOT_MODIFIED;
+            } else {
+                // Handle 'if-modified-since' header, only if 'if-none-match' does not exist.
+                // ie return here, because If-None-Match is definitive.
+                return PERFORM_METHOD;
+            }
+        }
+
+        if (dataLastModified != null) {
+            ETagResponse ifModifiedSinceResponse = ifModifiedSince(reqHeaders, dataLastModified);
+            if (ifModifiedSinceResponse != PERFORM_METHOD) {
+                return ifModifiedSinceResponse;
+            }
+            return ifUnmodifiedSince(reqHeaders, dataLastModified);
+        }
+        return PERFORM_METHOD;
```

Comment:
It seems like we have to return different statuses(304, 412, 206) depending on the evaluation of the precondition.
Instead of using `ETagResponse`, how about just returning an `HttpStatus` if a condition is not met? We can just return `null` if all conditions are met, so that the service can perform the request action.